pkg/tenant/tenants.go

```diff
@@ -139,6 +138,62 @@ func (s *Server) TenantDelete(c *gin.Context) {
 	c.Status(http.StatusOK)
 }
 
+// TenantUpdate will update a tenants records and
+// returns a 200 OK response.
+//
+// Route: /tenant/:tenantID
 func (s *Server) TenantUpdate(c *gin.Context) {
-	c.JSON(http.StatusNotImplemented, "not implemented yet")
+	var (
+		err    error
+		tenant *api.Tenant
+	)
+
+	// TODO: authentication and authorization middleware
+
+	// Bind the user request with JSON and return a 400 response if
+	// binding is not successful.
+	if err = c.BindJSON(&tenant); err != nil {
+		log.Warn().Err(err).Msg("could not parse tenant update request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind request"))
+		return
+	}
+
+	// Get the tenant ID from the URL and return a 400 if the tenant
+	// does not exist.
+	var tenantID ulid.ULID
+	if tenantID, err = ulid.Parse(c.Param("tenantID")); err != nil {
+		log.Debug().Err(err).Msg("could not parse tenant ulid")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse tenant id"))
+		return
+	}
+
+	// Verify the tenant name exists and return a 400 response if it does not exist.
+	if tenant.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("tenant name is required"))
+		return
+	}
+
+	// Verify the tenant environment type exists and return a 400 response if it does
+	// not exist.
+	if tenant.EnvironmentType == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("tenant environment type is required"))
+		return
+	}
+
+	// Prepare update request for insertion into the database.
+	req := &db.Tenant{
+		ID:              tenantID,
+		Name:            tenant.Name,
+		EnvironmentType: tenant.EnvironmentType,
+	}
```

Comment:
The pattern we've mostly used in GDS for Update handlers is to first try to retrieve the original record, then update the record that was retrieved and store it back to the database.

I think this would go a long way to make this endpoint more resilient to future changes, since it prevents us from accidentally overwriting internal data on the tenant in the database. We may also have metadata fields like `created_at`, `modified_at` timestamps that we don't want to override. This would also allow us to handle the "tenant not found" and "error calling the database" errors separately to give the caller more information about what went wrong.

Reply:
@pdeziel I made an update to retrieve the tenant from the db once the tenant ID is parsed from the URL. However, if the change isn't in line with what you had in mind please let me know.

Reply:
Great, now that we are retrieving the tenant from the db can we use that instead of the `req` (so modifying the model directly) when storing to the database?

Also, I think it's probably better to put the `db.RetrieveTenant` after the 400 checks, so that we don't have to make a request to the database unless the request is formatted correctly (this will save a bit of time on the error paths).